File: apps/desktop/src/lib/assets.ts
/**
 * Pure helpers for asset management — snippet generation, preamble detection,
 * and file-type classification. All functions are I/O-free and fully testable.
 */

/** Extensions that can be previewed as images. */
const IMAGE_EXTS: ReadonlySet<string> = new Set([
  'png',
  'jpg',
  'jpeg',
  'gif',
  'bmp',
  'svg',
  'tif',
  'tiff'
]);

/**
 * Return `true` when `filename` has an image extension that can be previewed
 * inline. Case-insensitive.
 */
export function isImageExt(filename: string): boolean {
  const dot = filename.lastIndexOf('.');
  if (dot < 0) {
    return false;
  }
  return IMAGE_EXTS.has(filename.slice(dot + 1).toLowerCase());
}

/**
 * Generate the LaTeX `\includegraphics` snippet for inserting `relPath` at
 * the cursor.
 *
 * @example
 *   assetSnippet('assets/fig.png')
 *   // → '\\includegraphics[width=\\linewidth]{assets/fig.png}'
 */
export function assetSnippet(relPath: string): string {
  return `\\includegraphics[width=\\linewidth]{${relPath}}`;
}

/**
 * Return `true` when `source` contains `\includegraphics` but not
 * `\graphicspath` — the cue to offer adding `\graphicspath{{assets/}}`.
 */
export function needsGraphicspath(source: string): boolean {
  return source.includes('\\includegraphics') && !source.includes('\\graphicspath');
}

/**
 * Insert `\graphicspath{{assets/}}` into `source` before `\begin{document}`.
 * When `\begin{document}` is absent the directive is appended at the end.
 *
 * Assumes `source` does not already contain `\graphicspath`.
 */
export function insertGraphicspath(source: string): string {
  const marker = '\\begin{document}';
  const idx = source.indexOf(marker);
  if (idx < 0) {
    return source + '\n\\graphicspath{{assets/}}\n';
  }
  return source.slice(0, idx) + '\\graphicspath{{assets/}}\n' + source.slice(idx);
}

File: crates/galley-core/src/assets.rs
//! Asset-related pure helpers: snippet generation and preamble detection.

/// Generate the LaTeX snippet for inserting an image at `rel_path`.
///
/// Produces a bare `\includegraphics` command so the caller can paste it at
/// the cursor. The path is project-relative (e.g. `assets/figure.png`).
pub fn figure_snippet(rel_path: &str) -> String {
    format!(r"\includegraphics[width=\linewidth]{{{rel_path}}}")
}

/// Return `true` when `source` contains `\includegraphics` but not
/// `\graphicspath` — the cue to offer adding `\graphicspath{{assets/}}`.
pub fn needs_graphicspath(source: &str) -> bool {
    source.contains(r"\includegraphics") && !source.contains(r"\graphicspath")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn figure_snippet_wraps_rel_path() {
        assert_eq!(
            figure_snippet("assets/diagram.png"),
            r"\includegraphics[width=\linewidth]{assets/diagram.png}"
        );
    }

    #[test]
    fn figure_snippet_with_subdirectory_path() {
        assert_eq!(
            figure_snippet("assets/figs/plot.pdf"),
            r"\includegraphics[width=\linewidth]{assets/figs/plot.pdf}"
        );
    }

    #[test]
    fn needs_graphicspath_false_when_no_includegraphics() {
        assert!(!needs_graphicspath(r"\documentclass{article}"));
    }

    #[test]
    fn needs_graphicspath_true_when_includegraphics_without_path() {
        assert!(needs_graphicspath(
            r"\documentclass{article}\n\includegraphics[width=\linewidth]{fig.png}"
        ));
    }

    #[test]
    fn needs_graphicspath_false_when_graphicspath_already_set() {
        assert!(!needs_graphicspath(
            r"\graphicspath{{assets/}}\n\includegraphics{fig.png}"
        ));
    }
}
